<html>
          <head>
            <title>Artificial Immune Recognition System | Clever Algorithms</title><link href="https://cleveralgorithms.com/nature-inspired/clever.css" media="screen" rel="stylesheet" type="text/css"/>
          <!-- support to display ruby source nicely -->
            <link href="https://cleveralgorithms.com/nature-inspired/prettify.css" type="text/css" rel="stylesheet"/>
            <script type="text/javascript" src="https://cleveralgorithms.com/nature-inspired/prettify.js"></script>
          </head>
        <!-- call to display ruby source nicely -->
        <body onload="prettyPrint()"><div class="container"><!-- Start Header -->
          <center>
          <h1>Clever Algorithms: Nature-Inspired Programming Recipes</h1>
          <em>A book by Jason Brownlee</em>

          <p>
            <a href="/">Home</a> |
            <a href="https://amzn.to/4iKM9uc">Amazon</a> |
            <a href="https://www.goodreads.com/book/show/10321060-clever-algorithms">GoodReads</a> |
            <a href="https://www.google.com.au/books/edition/Clever_Algorithms/SESWXQphCUkC">Google Books</a> |
            <a href="https://raw.githubusercontent.com/Jason2Brownlee/CleverAlgorithms/master/release/clever_algorithms.pdf">PDF</a> (<a href="https://raw.githubusercontent.com/Jason2Brownlee/CleverAlgorithms/master/release/clever_algorithms-src.zip">code</a>) |
            <a href="https://github.com/Jason2Brownlee/CleverAlgorithms">GitHub</a>
          </p>

          </center>
          <hr/>
          <br/>
          <!-- End Header --><div class='breadcrumb'>
<a href='../index.html'>Table of Contents</a>
&gt;&gt;
<a href='../immune.html'>Immune Algorithms</a>
&gt;&gt;
<a href='airs.html'>Artificial Immune Recognition System</a>
</div>
<h1><a name='artificial_immune_recognition_system'>Artificial Immune Recognition System</a></h1>
<p>
<em>Artificial Immune Recognition System, AIRS.</em>
</p>

<h2><a name='taxonomy'>Taxonomy</a></h2>
<p>
The Artificial Immune Recognition System belongs to the field of Artificial Immune Systems, and more broadly to the field of Computational Intelligence.
It was extended early to the canonical version called the  Artificial Immune Recognition System 2 (AIRS2) and provides the basis for extensions such as the Parallel Artificial Immune Recognition System  [<a href='#Watkins2004'>Watkins2004</a>].
It is related to other Artificial Immune System algorithms such as the Dendritic Cell Algorithm, the Clonal Selection Algorithm, and the Negative Selection Algorithm.
</p>


<h2><a name='inspiration'>Inspiration</a></h2>
<p>
The Artificial Immune Recognition System is inspired by the Clonal Selection theory of acquired immunity.
The clonal selection theory credited to Burnet was proposed to account for the behavior and capabilities of antibodies in the acquired immune system  [<a href='#Burnet1957'>Burnet1957</a>] [<a href='#Burnet1959'>Burnet1959</a>]. Inspired itself by the principles of Darwinian natural selection theory of evolution, the theory proposes that antigens select-for lymphocytes (both B and T-cells). When a lymphocyte is selected and binds to an antigenic determinant, the cell proliferates making many thousands more copies of itself and differentiates into different cell types (plasma and memory cells). Plasma cells have a short lifespan and produce vast quantities of antibody molecules, whereas memory cells live for an extended period in the host anticipating future recognition of the same determinant. The important feature of the theory is that when a cell is selected and proliferates, it is subjected to small copying errors (changes to the genome called somatic hypermutation) that change the shape of the expressed receptors. It also affects the  subsequent determinant recognition capabilities of both the antibodies bound to the lymphocytes cells surface, and the antibodies that plasma cells produce.
</p>


<h2><a name='metaphor'>Metaphor</a></h2>
<p>
The theory suggests that starting with an initial repertoire of general immune cells, the system is able to change itself (the compositions and densities of cells and their receptors) in response to experience with the environment. Through a blind process of selection and accumulated variation on the large scale of many billions of cells, the acquired immune system is capable of acquiring the necessary information to protect the host organism from the specific pathogenic dangers of the environment. It also suggests that the system must anticipate (guess) at the pathogen to which it will be exposed, and requires exposure to pathogen that may harm the host before it can acquire the necessary information to provide a defense.
</p>


<h2><a name='strategy'>Strategy</a></h2>
<p>
The information processing objective of the technique is to prepare a set of real-valued vectors to classify patterns.
The Artificial Immune Recognition System maintains a pool of memory cells that are prepared by exposing the system to a single iteration of the training data. Candidate memory cells are prepared when the memory cells are insufficiently stimulated for a given input pattern. A process of cloning and mutation of cells occurs for the most stimulated memory cell. The clones compete with each other for entry into the memory pool based on stimulation and on the amount of resources each cell is using. This concept of resources comes from prior work on Artificial Immune Networks, where a single cell (an Artificial Recognition Ball or ARB) represents a set of similar cells. Here, a cell's resources are a function of its stimulation to a given input pattern and the number of clones it may create.
</p>


<h2><a name='procedure'>Procedure</a></h2>
<p>
Algorithm (below) provides a high-level pseudocode for preparing memory cell vectors using the Artificial Immune Recognition System, specifically the canonical AIRS2.
An affinity (distance) measure between input patterns must be defined. For real-valued vectors, this is commonly the Euclidean distance:
</p>
$dist(x,c) = \sum_{i=1}^{n} (x_i - c_i)^2$
<p>
where $["n"]$ is the number of attributes, $["x"]$ is the input vector and $["c"]$ is a given cell vector. The variation of cells during cloning (somatic hypermutation) occurs inversely proportional to the stimulation of a given cell to an input pattern.
</p>
<div class='pseudocode'>
<strong><strong><code>Input</code></strong></strong>: 
<code>InputPatterns</code>, $clone_{rate}$, $mutate_{rate}$, $stim_{thresh}$, $resources_{max}$, $affinity_{thresh}$
<br />
<strong><strong><code>Output</code></strong></strong>: 
$Cells_{memory}$
<br />
$Cells_{memory}$ $["\\leftarrow"]$ <code>InitializeMemoryPool</code>(<code>InputPatterns</code>)<br />
<strong><code>For</code></strong> ($InputPattern_i$ $["\\in"]$ <code>InputPatterns</code>)<br />
&nbsp;&nbsp;&nbsp;&nbsp;<code>Stimulate</code>($Cells_{memory}$, <code>InputPatterns</code>)<br />
&nbsp;&nbsp;&nbsp;&nbsp;$Cell_{best}$ $["\\leftarrow"]$ <code>GetMostStimulated</code>($InputPattern_i$, $Cells_{memory}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>If</code></strong> ($Cell_{best}^{class}$ $["\\neq"]$ $InputPattern_{i}^{class}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cells_{memory}$ $["\\leftarrow"]$ <code>CreateNewMemoryCell</code>($InputPattern_i$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>Else</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Clones_{num}$ $["\\leftarrow"]$ $Cell_{best}^{stim}$ $["\\times"]$ $clone_{rate}$ $["\\times"]$ $mutate_{rate}$<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cells_{clones}$ $["\\leftarrow"]$ $Cell_{best}$<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>For</code></strong> ($["i"]$ <strong><code>To</code></strong> $Clones_{num}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cells_{clones}$ $["\\leftarrow"]$ <code>CloneAndMutate</code>($Cell_{best}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>While</code></strong> (<code>AverageStimulation</code>($Cells_{clones}$) $["\\leq"]$ $stim_{thresh}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>For</code></strong> ($Cell_{i}$ $["\\in"]$ $Cells_{clones}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cells_{clones}$ $["\\leftarrow"]$ <code>CloneAndMutate</code>($Cell_{i}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>Stimulate</code>($Cells_{clones}$, <code>InputPatterns</code>)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>ReducePoolToMaximumResources</code>($Cells_{clones}$, $resources_{max}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cell_{c}$ $["\\leftarrow"]$ <code>GetMostStimulated</code>($InputPattern_i$, $Cells_{clones}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>If</code></strong> ($Cell_{c}^{stim}$ &gt; $Cell_{best}^{stim}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;$Cells_{memory}$ $["\\leftarrow"]$ $Cell_{c}$<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>If</code></strong> (<code>Affinity</code>($Cell_{c}$, $Cell_{best}$) $["\\leq"]$ $affinity_{thresh}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<code>DeleteCell</code>($Cell_{best}$, $Cells_{memory}$)<br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
&nbsp;&nbsp;&nbsp;&nbsp;<strong><code>End</code></strong><br />
<strong><code>End</code></strong><br />
<strong><code>Return</code></strong> ($Cells_{memory}$)<br />
</div>
<div class='caption'>Pseudocode for AIRS2.</div>



<h2><a name='heuristics'>Heuristics</a></h2>
<ul>
<li> The AIRS was designed as a supervised algorithm for classification problem domains.</li>
<li> The AIRS is non-parametric, meaning that it does not rely on assumptions about that structure of the function that is is approximating.</li>
<li> Real-values in input vectors should be normalized such that $["x \\in [0,1)"]$.</li>
<li> Euclidean distance is commonly used to measure the distance between real-valued vectors (affinity calculation), although other distance measures may be used (such as dot product), and data specific distance measures may be required for non-scalar attributes.</li>
<li> Cells may be initialized with small random values or more commonly with values from instances in the training set.</li>
<li> A cell's affinity is typically minimizing, where as a cells stimulation is maximizing and typically $["\\in [0,1]"]$.</li>
</ul>


<h2><a name='code_listing'>Code Listing</a></h2>
<p>
Listing (below) provides an example of the Artificial Immune Recognition System implemented in the Ruby Programming Language.
The problem is a contrived classification problem in a 2-dimensional domain $["x\\in[0,1], y\\in[0,1]"]$ with two classes: 'A' ($["x\\in[0,0.4999999], y\\in[0,0.4999999]"]$) and 'B' ($["x\\in[0.5,1], y\\in[0.5,1]"]$).
</p>
<p>
The algorithm is an implementation of the AIRS2 algorithm  [<a href='#Watkins2002b'>Watkins2002b</a>]. An initial pool of memory cells is created, one cell for each class. Euclidean distance divided by the maximum possible distance in the domain is taken as the affinity and stimulation is taken as $["1.0-affinity"]$. The meta-dynamics for memory cells (competition for input patterns) is not performed and may be added into the implementation as an extension.
</p>
<pre class='prettyprint lang-rb'>
def random_vector(minmax)
  return Array.new(minmax.size) do |i|
    minmax[i][0] + ((minmax[i][1] - minmax[i][0]) * rand())
  end
end

def generate_random_pattern(domain)
  class_label = domain.keys[rand(domain.keys.size)]
  pattern = {:label=&gt;class_label}
  pattern[:vector] = random_vector(domain[class_label])
  return pattern
end

def create_cell(vector, class_label)
  return {:label=&gt;class_label, :vector=&gt;vector}
end

def initialize_cells(domain)
  mem_cells = []
  domain.keys.each do |key|
    mem_cells &lt;&lt; create_cell(random_vector([[0,1],[0,1]]), key)
  end
  return mem_cells
end

def distance(c1, c2)
  sum = 0.0
  c1.each_index {|i| sum += (c1[i]-c2[i])**2.0}
  return Math.sqrt(sum)
end

def stimulate(cells, pattern)
  max_dist = distance([0.0,0.0], [1.0,1.0])
  cells.each do |cell|
    cell[:affinity] = distance(cell[:vector], pattern[:vector]) / max_dist
    cell[:stimulation] = 1.0 - cell[:affinity]
  end
end

def get_most_stimulated_cell(mem_cells, pattern)
  stimulate(mem_cells, pattern)
  return mem_cells.sort{|x,y| y[:stimulation] &lt;=&gt; x[:stimulation]}.first
end

def mutate_cell(cell, best_match)
  range = 1.0 - best_match[:stimulation]
  cell[:vector].each_with_index do |v,i|
    min = [(v-(range/2.0)), 0.0].max
    max = [(v+(range/2.0)), 1.0].min
    cell[:vector][i] = min + (rand() * (max-min))
  end
  return cell
end

def create_arb_pool(pattern, best_match, clone_rate, mutate_rate)
  pool = []
  pool &lt;&lt; create_cell(best_match[:vector], best_match[:label])
  num_clones = (best_match[:stimulation] * clone_rate * mutate_rate).round
  num_clones.times do
    cell = create_cell(best_match[:vector], best_match[:label])
    pool &lt;&lt; mutate_cell(cell, best_match)
  end
  return pool
end

def competition_for_resournces(pool, clone_rate, max_res)
  pool.each {|cell| cell[:resources] = cell[:stimulation] * clone_rate}
  pool.sort!{|x,y| x[:resources] &lt;=&gt; y[:resources]}
  total_resources = pool.inject(0.0){|sum,cell| sum + cell[:resources]}
  while total_resources &gt; max_res
    cell = pool.delete_at(pool.size-1)
    total_resources -= cell[:resources]
  end
end

def refine_arb_pool(pool, pattern, stim_thresh, clone_rate, max_res)
  mean_stim, candidate = 0.0, nil
  begin
    stimulate(pool, pattern)
    candidate = pool.sort{|x,y| y[:stimulation] &lt;=&gt; x[:stimulation]}.first
    mean_stim = pool.inject(0.0){|s,c| s + c[:stimulation]} / pool.size
    if mean_stim &lt; stim_thresh
      candidate = competition_for_resournces(pool, clone_rate, max_res)
      pool.size.times do |i|
        cell = create_cell(pool[i][:vector], pool[i][:label])
        mutate_cell(cell, pool[i])
        pool &lt;&lt; cell
      end
    end
  end until mean_stim &gt;= stim_thresh
  return candidate
end

def add_candidate_to_memory_pool(candidate, best_match, mem_cells)
  if candidate[:stimulation] &gt; best_match[:stimulation]
    mem_cells &lt;&lt; candidate
  end
end

def classify_pattern(mem_cells, pattern)
  stimulate(mem_cells, pattern)
  return mem_cells.sort{|x,y| y[:stimulation] &lt;=&gt; x[:stimulation]}.first
end

def train_system(mem_cells, domain, num_patterns, clone_rate, mutate_rate, stim_thresh, max_res)
  num_patterns.times do |i|
    pattern = generate_random_pattern(domain)
    best_match = get_most_stimulated_cell(mem_cells, pattern)
    if best_match[:label] != pattern[:label]
      mem_cells &lt;&lt; create_cell(pattern[:vector], pattern[:label])
    elsif best_match[:stimulation] &lt; 1.0
      pool = create_arb_pool(pattern, best_match, clone_rate, mutate_rate)
      cand = refine_arb_pool(pool,pattern, stim_thresh, clone_rate, max_res)
      add_candidate_to_memory_pool(cand, best_match, mem_cells)
    end
    puts " &gt; iter=#{i+1}, mem_cells=#{mem_cells.size}"
  end
end

def test_system(mem_cells, domain, num_trials=50)
  correct = 0
  num_trials.times do
    pattern = generate_random_pattern(domain)
    best = classify_pattern(mem_cells, pattern)
    correct += 1 if best[:label] == pattern[:label]
  end
  puts "Finished test with a score of #{correct}/#{num_trials}"
  return correct
end

def execute(domain, num_patterns, clone_rate, mutate_rate, stim_thresh, max_res)
  mem_cells = initialize_cells(domain)
  train_system(mem_cells, domain, num_patterns, clone_rate, mutate_rate, stim_thresh, max_res)
  test_system(mem_cells, domain)
  return mem_cells
end

if __FILE__ == $0
  # problem configuration
  domain = {"A"=&gt;[[0,0.4999999],[0,0.4999999]],"B"=&gt;[[0.5,1],[0.5,1]]}
  num_patterns = 50
  # algorithm configuration
  clone_rate = 10
  mutate_rate = 2.0
  stim_thresh = 0.9
  max_res = 150
  # execute the algorithm
  execute(domain, num_patterns, clone_rate, mutate_rate, stim_thresh, max_res)
end
</pre>
<div class='caption'>AIRS in Ruby</div>
<div class='download_src'>Download: <a href='airs.rb'>airs.rb</a>.</div>


<h2><a name='references'>References</a></h2>

<h3><a name='primary_sources'>Primary Sources</a></h3>
<p>
The Artificial Immune Recognition System was proposed in the Masters work by Watkins  [<a href='#Watkins2001'>Watkins2001</a>], and later published  [<a href='#Watkins2002a'>Watkins2002a</a>].
Early works included the application of the AIRS by Watkins and Boggess to a suite of benchmark classification problems  [<a href='#Watkins2002'>Watkins2002</a>], and a similar study by Goodman and Boggess comparing to a conceptually similar approach called Learning Vector Quantization  [<a href='#Goodman2002'>Goodman2002</a>].
</p>


<h3><a name='learn_more'>Learn More</a></h3>
<p>
Marwah and Boggess investigated the algorithm seeking issues that affect the algorithms performance  [<a href='#Marwah2002'>Marwah2002</a>]. They compared various variations of the algorithm with modified resource allocation schemes, tie-handling within the ARB pool, and ARB pool organization.
Watkins and Timmis proposed a new version of the algorithm called AIRS2 which became the replacement for AIRS1  [<a href='#Watkins2002b'>Watkins2002b</a>]. The updates reduced the complexity of the approach while maintaining the accuracy of the results. An investigation by Goodman et al. into the so called '<em>source of power</em>' in AIRS indicated that perhaps the memory cell maintenance procedures played an important role  [<a href='#Goodman2003'>Goodman2003</a>].
Watkins et al. provide a detailed review of the technique and its application  [<a href='#Watkins2004a'>Watkins2004a</a>].
</p>




<h2>Bibliography</h2>
<table>
 <tr valign="top">
 <td><a name='Burnet1957'>[Burnet1957]</a></td>
 <td>["F. M. Burnet"], "<a href="http://scholar.google.com.au/scholar?q=["A+modification+of+Jerne's+theory+of+antibody+production+using+the\n\tconcept+of+clonal+selection"]">["A modification of Jerne's theory of antibody production using the\n\tconcept of clonal selection"]</a>", ["Australian Journal of Science"], ["1957"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Burnet1959'>[Burnet1959]</a></td>
 <td>["F. M. Burnet"], "<a href="http://scholar.google.com.au/scholar?q=["The+clonal+selection+theory+of+acquired+immunity"]">["The clonal selection theory of acquired immunity"]</a>", ["Vanderbilt University Press"], ["1959"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Goodman2002'>[Goodman2002]</a></td>
 <td>D. E. ["Goodman Jr. and L. Boggess and A."] Watkins, "<a href="http://scholar.google.com.au/scholar?q=["Artificial+Immune+System+Classification+of+Multiple-Class+Problems"]">["Artificial Immune System Classification of Multiple-Class Problems"]</a>", in ["Fuzzy Logic, Evolutionary Programming, Complex Systems and Artificial\n\tLife"], ["2002"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Goodman2003'>[Goodman2003]</a></td>
 <td>D. E. ["Goodman Jr. and L. Boggess and A."] Watkins, "<a href="http://scholar.google.com.au/scholar?q=["An+Investigation+into+the+Source+of+Power+for+AIRS,+an+Artificial\n\tImmune+Classification+Systems"]">["An Investigation into the Source of Power for AIRS, an Artificial\n\tImmune Classification Systems"]</a>", in ["Proceedings of the International Joint Conference on Neural Networks\n\t(IJCNN'03)"], ["2003"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Marwah2002'>[Marwah2002]</a></td>
 <td>G. ["Marwah and L."] Boggess, "<a href="http://scholar.google.com.au/scholar?q=["Artificial+Immune+Systems+for+classification:+Some+issues"]">["Artificial Immune Systems for classification: Some issues"]</a>", in ["First International Conference on Artificial Immune Systems"], ["2002"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2001'>[Watkins2001]</a></td>
 <td>A. B."] ["Watkins, "<a href="http://scholar.google.com.au/scholar?q=["AIRS:+A+resource+limited+artificial+immune+classifiers"]">["AIRS: A resource limited artificial immune classifiers"]</a>", [Masters Thesis] ["Mississippi State University"], ["2001"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2002'>[Watkins2002]</a></td>
 <td>A. ["Watkins and L."] Boggess, "<a href="http://scholar.google.com.au/scholar?q=["A+New+Classifier+Based+on+Resource+Limited+Artificial+Immune+Systems"]">["A New Classifier Based on Resource Limited Artificial Immune Systems"]</a>", in ["Part of the 2002 IEEE World Congress on Computational Intelligence"], ["2002"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2002a'>[Watkins2002a]</a></td>
 <td>A. B. ["Watkins and L. C."] Boggess, "<a href="http://scholar.google.com.au/scholar?q=["A+Resource+Limited+Artificial+Immune+Classifier"]">["A Resource Limited Artificial Immune Classifier"]</a>", in ["Part of the 2002 IEEE World Congress on Computational Intelligence\n\theld in Honolulu"], ["2002"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2002b'>[Watkins2002b]</a></td>
 <td>A. ["Watkins and J."] Timmis, "<a href="http://scholar.google.com.au/scholar?q=["Artificial+Immune+Recognition+System+(AIRS):+Revisions+and+Refinements"]">["Artificial Immune Recognition System (AIRS): Revisions and Refinements"]</a>", in ["1st International Conference on Artificial Immune Systems (ICARIS2002)"], ["2002"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2004'>[Watkins2004]</a></td>
 <td>A. ["Watkins and J."] Timmis, "<a href="http://scholar.google.com.au/scholar?q=["Exploiting+Parallelism+Inherent+in+AIRS,+an+Artificial+Immune+Classifier"]">["Exploiting Parallelism Inherent in AIRS, an Artificial Immune Classifier"]</a>", in ["Lecture Notes in Computer Science (LNCS)"], ["2004"].</td>
 </tr>
 <tr valign="top">
 <td><a name='Watkins2004a'>[Watkins2004a]</a></td>
 <td>A. ["Watkins and J. Timmis and L."] Boggess, "<a href="http://scholar.google.com.au/scholar?q=["Artificial+Immune+Recognition+System+(AIRS):+An+Immune-Inspired\n\tSupervised+Learning+Algorithms"]">["Artificial Immune Recognition System (AIRS): An Immune-Inspired\n\tSupervised Learning Algorithms"]</a>", ["Genetic Programming and Evolvable Machines"], ["2004"].</td>
 </tr>
</table>

<!-- generate math images from latex math snippets -->
    <script type="text/javascript" src="https://mathcache.s3.amazonaws.com/replacemath.js"></script>
    <script type="text/javascript">replaceMath(document.body);</script>
        </div></body>
          </html>